### DISCUSSION TITLE
OutputMixer issue

### DISCUSSION THREAD

**Discussion by solonello** [pschatzmann/arduino-audio-tools#1348] [Q&A] [answered]
Hi!

I'm trying to use the output mixer to mix two audio sources. The first one is from the BluetoothA2DPSink (channel 0), the second one from some audio files stored on an SD card that are played based on the click of some buttons (channel 1).
Being inspired from another thread, I created the output mixer:

```OutputMixer<int16_t> audioMixer(i2sStream, 2);```

Then inside the stream reader callback of the BluetoothA2DPSink I do something like this:

```
void onBluetoothDataRead(const uint8_t *data, uint32_t length)
{
    int open = length;
    int pos = 0;

    while (open > 0)
    {
        int writeSize = min(open, 1024);
        audioMixer.write(0, data + pos, writeSize);

        audioMixer.setWeight(0, audioMixer.availableForRead(0) ? 1 : 0);
        audioMixer.setWeight(1, audioMixer.availableForRead(1) ? 1 : 0);
        audioMixer.flushMixer();

        open -= writeSize;
        pos += writeSize;
    }
}
```
availableForRead(channel) is a custom function that checks if the provided channel (buffer of the OutputMixer) contains data.
The channel 1 is written similarly, inside the loop function, by reading from an EncodedAudioStream object.

Right after the first flush of the mixer, I get this warning and of course no actual sound is played:
`[W] AudioOutput.h : 371 - Available Buffer too small 0: requested: 512 -> increase the buffer size`

Any ideas of why is this happening or suggestions on how to improve this logic?
Thanks

**Comment by pschatzmann** [comment, marked as answer]
Just try to increase the copy buffer size that it defined in the begin of the output mixer

**Comment by solonello** [comment]
Hi Phil, thanks for your answer. I begin the mixer like this:

`audioMixer.begin(1024, RAM);`

Shouldn't it be enough since the amount of data copied on each iteration is 1024 as well?
I'm asking just to try to better understand how all of this "stuff" work, in the afternoon I will try to increase the buffer size as you suggested.
Thanks again!
